src/vscode-bicep/src/feedback/surveys.ts

```diff
@@ -0,0 +1,338 @@
+import { callWithTelemetryAndErrorHandling } from "@microsoft/vscode-azext-utils";
+import assert from "assert";
+import { GlobalState, GlobalStateKeys } from "../globalState";
+import https from "https";
+import { daysToMs, monthsToDays } from "../utils/time";
+
+// Annual HaTS survey
+const hatsAnnualSurveyInfo: ISurveyInfo = {
+  akaLinkToSurvey: "bicep/surveys/annual",
+  // Enough to be sure we don't ask again before the next survey, but still have flexibility about sending out the next
+  //   survey earlier in the year if we want to.
+  postponeAfterTakenInDays: monthsToDays(6),
+  surveyPrompt:
+    "Could you please take 2 minutes to tell us how well Bicep is working for you?",
+  postponeForLaterInDays: 1 * 7,
+  surveyStateKey: GlobalStateKeys.annualSurveyStateKey,
+};
+
+export function showSurveys(globalState: GlobalState): void {
+  checkShowSurvey(new Survey(globalState, hatsAnnualSurveyInfo));
+}
+
+export function checkShowSurvey(survey: Survey): void {
+  // Don't wait
+  callWithTelemetryAndErrorHandling(
+    "survey",
+    async (context: IActionContext) => {
+      await survey.checkShowSurvey(context, new Date());
+    }
+  );
+}
+
+export interface ISurveyInfo {
+  akaLinkToSurvey: string; // the part after https://aka.ms/
+  surveyStateKey: string; // Key used in global storage to persist the survey state
+  postponeAfterTakenInDays: number;
+  postponeForLaterInDays: number;
+  surveyPrompt: string;
+}
+
+export interface ISurveyState {
+  lastTaken?: Date;
+  postponedUntil?: Date;
+}
+
+export interface IPersistedSurveyState {
+  lastTakenMs?: number;
+  postponedUntilMs?: number;
+}
+
+export class Survey {
+  public constructor(
+    private readonly globalState: GlobalState,
+    private readonly surveyInfo: ISurveyInfo,
+    private inject: {
+      showInformationMessage: typeof window.showInformationMessage;
+      getIsSurveyAvailable: typeof Survey.getIsSurveyAvailable;
+      launchSurvey: typeof Survey.launchSurvey;
+    } = {
+        showInformationMessage: window.showInformationMessage,
+        getIsSurveyAvailable: Survey.getIsSurveyAvailable,
+        launchSurvey: Survey.launchSurvey,
+      }
+  ) {
+    // noop
+  }
+
+  /**
+   * Shows the survey if it's available and timely, and the user doesn't opt out.
+   */
+  public async checkShowSurvey(
+    context: IActionContext,
+    now: Date
+  ): Promise<void> {
+    context.errorHandling.suppressDisplay = true;
+    context.telemetry.properties.isActivationEvent = "true";
+    context.telemetry.properties.akaLink = this.surveyInfo.akaLinkToSurvey;
+
+    const surveyState = this.getPersistedSurveyState(context, now);
+
+    const shouldShowSurvey = await this.shouldAskToTakeSurvey(
+      context,
+      surveyState,
+      now
+    );
+
+    if (shouldShowSurvey) {
+      await this.askToTakeSurvey(context, surveyState, now);
+    }
+
+    await this.updatePersistedSurveyState(context, surveyState);
```

Comment:
This would only handle the case when user clicked on one of the three options: Never ask again, Maybe later, sure. 

How are we handling this case?
```
If they ignore the popup, wait this long before considering asking again: 12 weeks
```

Reply:
I debugged through the code and looks like in case of "ignore" we show the notification again. Is this intentional?

Reply:
Sorry, I'm not implementing the algorithm mentioned in the bug (turns out that's not what Alex intended for this survey).  The current algorithm is simply to check every time the extension starts up.  So if they ignore they'll see it again the next time (or not see it, as the case may be) until they actually click one of the buttons, yeah.  Seems reasonable to me, do you disagree?

Reply:
Thanks for the explanation. Could we please update the PR description with the rules? In case of ignore, looks like it'll be sent again during the next language server startup session. If this is intentional, then I am okay.

Reply:
Already updated the bug:

> Not implementing the algorithm above, but this simpler one:
> 
> The current algorithm is simply to check every time the extension starts up. So if they ignore they'll see it again the next time they start up vs code and bicep, until they actually click one of the buttons, and then not again until the survey is available again (for the current survey, once per year). Everybody gets asked at the same time, no randomization or trying to stretch out the time for the survey.